packages/react-native-reanimated/src/hook/useComposedEventHandler.ts

```diff
@@ -26,55 +25,239 @@ export function useComposedEventHandler<
+  const workletsRecord: Record<string, WorkletFunction> = {};
 
-  handlers
+  const workletHandlers = handlers
     .filter((h) => h !== null)
-    .forEach((handler) => {
-      // EventHandlerProcessed is the return type of useEvent and has to be force casted to EventHandlerInternal, because we need WorkletEventHandler object
-      const { workletEventHandler } =
-        handler as unknown as EventHandlerInternal<Context>;
-      if (workletEventHandler instanceof WorkletEventHandler) {
-        workletEventHandler.eventNames.forEach((eventName) => {
-          composedEventNames.add(eventName);
-
-          if (workletsMap[eventName]) {
-            workletsMap[eventName].push(workletEventHandler.worklet);
-          } else {
-            workletsMap[eventName] = [workletEventHandler.worklet];
-          }
-
-          const handlerName = eventName + `${workletsMap[eventName].length}`;
-          workletsRecord[handlerName] =
-            workletEventHandler.worklet as WorkletFunction;
-        });
+    .filter((h) => isWorkletEventHandler(h));
+
+  workletHandlers.forEach((handler) => {
+    // EventHandlerProcessed is the return type of useEvent and has to be force casted to EventHandlerInternal, because we need WorkletEventHandler object
+    const { workletEventHandler } =
+      handler as unknown as EventHandlerInternal<Context>;
+    if (workletEventHandler instanceof WorkletEventHandler) {
+      workletEventHandler.eventNames.forEach((eventName) => {
+        composedEventNames.add(eventName);
+
+        if (workletsMap[eventName]) {
+          workletsMap[eventName].push(workletEventHandler.worklet);
+        } else {
+          workletsMap[eventName] = [workletEventHandler.worklet];
+        }
+
+        const handlerName = eventName + `${workletsMap[eventName].length}`;
+        workletsRecord[handlerName] =
+          workletEventHandler.worklet as WorkletFunction;
+      });
+    }
+  });
+
+  const { doDependenciesDiffer } = useHandler(workletsRecord);
+
+  // Record of provided JS handlers, used for passing it further down to WorkletEventHandler
+  const JSHandlersRecord: Partial<
+    Record<JSScrollEventsOutput, JSHandler<Event>>
+  > = {};
+  // Helper map used for merging same-event JS handlers into JSHandlersRecord object
+  const JSHandlersMap: Partial<
+    Record<JSScrollEventsInput, JSHandler<Event>[]>
+  > = {};
+  // Ref to compare changes in JSHandlers
+  const JSHandlersRef = useRef<JSHandlersObject<Event>[] | null>(null);
+  // Flag that lets JS handlers get rebuilt too
+  let JSHandlersNeedRebuild = false;
+
+  const JSHandlers: JSHandlersObject<Event>[] = handlers
+    .filter((h) => h !== null)
+    .filter((h) => isJSHandler(h)) as JSHandlersObject<Event>[];
+
+  // Update/initialize the ref and determine whether JS handlers need rebuild or not
+  if (JSHandlersRef.current === null) {
+    JSHandlersRef.current = JSHandlers;
+  } else {
+    JSHandlersNeedRebuild = !areJSHandlersEqual(
```

Comment:
Is it possible to rely on `useEffect` and the dependency array instead of implementing it on your own?

Reply:
And since it is an advanced hook and not many people are gonna use it, it doesn't need some super-tuned performance upgrades. For now it calculates everything it needs from scratch on every render and I didn't notice any slowdown caused by it